### ISSUE TITLE
Blazor: Preview9 made HtmlRenderer & ComponentRenderedText internal

### ISSUE THREAD

**Issue by chanan** [dotnet/aspnetcore#13687]
Two Libraries (that I know of) use HtmlRenderer & ComponentRenderedText prior to previw 9. One is BlazorStyled: https://github.com/chanan/BlazorStyled and the other is Egil's testing library: https://github.com/egil/razor-components-testing-library

Both use it for the same reason, to get the content of ChildContent. It was used like so:

```csharp
private string RenderAsString()
{
    string result = string.Empty;
    try
    {
        ParameterView paramView = ParameterView.FromDictionary(new Dictionary<string, object>() { { "ChildContent", ChildContent } });
        using HtmlRenderer htmlRenderer = new HtmlRenderer(_emptyServiceProvider, NullLoggerFactory.Instance, _encoder);
        IEnumerable<string> tokens = GetResult(htmlRenderer.Dispatcher.InvokeAsync(() => htmlRenderer.RenderComponentAsync<TempComponent>(paramView)));
        result = string.Join("", tokens.ToArray());
    }
    catch
    {
        //ignored dont crash if can't get result
    }
    return result;
}

private IEnumerable<string> GetResult(Task<ComponentRenderedText> task)
{
    if (task.IsCompleted && task.Status == TaskStatus.RanToCompletion && !task.IsFaulted && !task.IsCanceled)
    {
        return task.Result.Tokens;
    }
    else
    {
        ExceptionDispatchInfo.Capture(task.Exception).Throw();
        throw new InvalidOperationException("We will never hit this line");
    }
}
```

However as of Preview9 that is no longer possible as HtmlRenderer & ComponentRenderedText are both internal - Please make them public (**Or provide another way to get the content of ChildContent**).

CC: @danroth27 @SteveSandersonMS @egil

**Comment by danroth27** [comment]
@javiercn

**Comment by egil** [comment]
I would like to add, that having an easy to use API, that can use/by default uses the same service provider as the renderer it is inside to turn a renderfragment into a string, will likely be a great addition to the Blazor library. I'm pretty sure there are going to be other devs out there who will need this feature but won't have the @SteveSandersonMS chops to create a custom renderer for their purpose either :-)

**Comment by akorchev** [comment]
Yep, this one is quite a bummer and rendered (pun intended) Radzen's Blazor design time not working.

**Comment by javiercn** [comment]
In the past the APIs for accessing the data for producing the HTML were internal (private protected if I recall) but now they are public, so you can do the same thing HTML Renderer does yourself.

https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs

Let me know if you run into any roadblocks.

Personally I would like to also bring this back on a future release, but I would like to make some API changes to improve performance.

**Comment by akorchev** [comment]
@javiercn it uses an internal struct https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderedText.cs

**Comment by akorchev** [comment]
And there is this warning:

```
warning BL0006: The types in 'Microsoft.AspNetCore.Components.RenderTree' are not recommended for use outside of the Blazor framework. These  type definitions will change in future releases.
```

**Comment by egil** [comment]
> In the past the APIs for accessing the data for producing the HTML were internal (private protected if I recall) but now they are public, so you can do the same thing HTML Renderer does yourself.
> 
> https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs
> 
> Let me know if you run into any roadblocks.
> 
> Personally I would like to also bring this back on a future release, but I would like to make some API changes to improve performance.

@javiercn its no problem copy-pasting your version of `HtmlRenderer.cs` and `ComponentRenderedText.cs`. But please let us know when you have updates to them. Would very much like to stay compatible/up-to-date. Perf improvements sounds good, and please take input on a possible API update :)

> And there is this warning:
> 
> ```
> warning BL0006: The types in 'Microsoft.AspNetCore.Components.RenderTree' are not recommended for use outside of the Blazor framework. These  type definitions will change in future releases.
> ```

@akorchev just suppress the warning. Its fair its there, since we are digging our feet into the low level APIs in Blazor.

This works for me. Create a `GlobalSuppressions.cs` in your project and add the following to it:

```csharp
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "BL0006:The types in 'Microsoft.AspNetCore.Components.RenderTree' are not recommended for use outside of the Blazor framework. These type definitions will change in future releases.",
    Justification = "I will take the chance",
    Scope = "namespaceanddescendants",
    Target = "<YOUR.NAMESPACE.HERE")]
```

**Comment by akorchev** [comment]
@javiercn well instead of using low-level Blazor API we could have just used the higher level and now internal HtmlRenderer ;) 

@javiercn What is the reasoning of having public low-level warning-producing API instead of public HtmlRenderer?

**Comment by javiercn** [comment]
@akorchev The API in its current form is not very good.

@danroth27 Given the interest on this API can we consider the appropriate improvements in HtmlRenderer to make it public in 3.1? I don't think we should release the API in its current form if we have time to make it performant before 3.1. It should take 2 days tops.

**Comment by SteveSandersonMS** [comment]
People can easily copy the relevant bits of the `HtmlRenderer` sources into their own projects. It's what I did for [this unit testing prototype](https://github.com/SteveSandersonMS/BlazorUnitTestingPrototype) and took hardly any time. I'm not sure there's a need to rush this into 3.1.

**Comment by javiercn** [comment]
@SteveSandersonMS If there's demand for it and prevents people using the pubternal API I think it brings goodness.

**Comment by akorchev** [comment]
@javiercn @SteveSandersonMS are there any planned changes in the `Microsoft.AspNetCore.Components.RenderTree` namespace before the official 3.0 release?

**Comment by SteveSandersonMS** [comment]
@akorchev No render tree API changes are expected between preview 9 and final 3.0.

**Comment by michalczerwinski** [comment]
What is status of this? Can we at this point (after 3.1) assume using GetFrames() is acceptable?

Not being able to access it is a big limitation for more advanced Blazor scenarios. Some clear recommendation would be appreciated.

**Comment by javiercn** [comment]
There has been a public API for this since at least 8.0

https://github.com/dotnet/aspnetcore/commit/afd72717c019eb2b8f4e56ad39d1898a6d333fbc
